### PR TITLE
[SPARK-28753][SQL] Dynamically reuse subqueries in AQE

### PR DESCRIPTION
**Pull request by maryannxue**
### What changes were proposed in this pull request?
This PR changes subquery reuse in Adaptive Query Execution from compile-time static reuse to execution-time dynamic reuse. This PR adds a `ReuseAdaptiveSubquery` rule that applies to a query stage after it is created and before it is executed. The new dynamic reuse enables subqueries to be reused across all different subquery levels.

### Why are the changes needed?
This is an improvement to the current subquery reuse in Adaptive Query Execution, which allows subquery reuse to happen in a lazy fashion as well as at different subquery levels.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Passed existing tests.

### CODE REVIEW

**Comment by maryannxue** [comment]
cc @cloud-fan @hvanhovell @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 1]
```diff
@@ -73,17 +74,19 @@ case class AdaptiveSparkPlanExec(
...
 
+  @transient private val ensureRequirements = EnsureRequirements(conf)
+
   // A list of physical plan rules to be applied before creation of query stages. The physical
   // plan should reach a final status of query stages (i.e., no more addition or removal of
   // Exchange nodes) after running these rules.
-  @transient private val queryStagePreparationRules: Seq[Rule[SparkPlan]] = Seq(
-    PlanAdaptiveSubqueries(subqueryMap),
-    EnsureRequirements(conf)
+  private def queryStagePreparationRules: Seq[Rule[SparkPlan]] = Seq(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: how about
```
@transient private val queryStagePreparationRules: Seq[Rule[SparkPlan]] = {
  preprocessingRules :+ EnsureRequirements(conf)
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala [hunk 2]
```diff
@@ -119,6 +122,7 @@ case class AdaptiveSparkPlanExec(
     if (isFinalPlan) {
       currentPhysicalPlan.execute()
     } else {
+      currentPhysicalPlan = applyPhysicalRules(currentPhysicalPlan, queryStagePreparationRules)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: why is this needed?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/PlanAdaptiveSubqueries.scala [hunk 3]
```diff
@@ -17,13 +17,15 @@
...
+import scala.collection.mutable
+
 import org.apache.spark.sql.catalyst.expressions
 import org.apache.spark.sql.catalyst.expressions.ListQuery
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.execution.{ExecSubqueryExpression, SparkPlan}
 
 case class PlanAdaptiveSubqueries(
-    subqueryMap: Map[Long, ExecSubqueryExpression]) extends Rule[SparkPlan] {
+    subqueryMap: mutable.HashMap[Long, ExecSubqueryExpression]) extends Rule[SparkPlan] {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/PlanAdaptiveSubqueries.scala]: nin: we can write `scala.collect.Map` here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala [hunk 4]
```diff
@@ -53,12 +56,13 @@ case class InsertAdaptiveSparkPlan(
...
         val subqueryMap = buildSubqueryMap(plan)
-        // Run preparation rules.
-        val preparations = AdaptiveSparkPlanExec.createQueryStagePreparationRules(
-          session.sessionState.conf, subqueryMap)
-        val newPlan = AdaptiveSparkPlanExec.applyPhysicalRules(plan, preparations)
+        val planSubqueriesRule = PlanAdaptiveSubqueries(subqueryMap)
+        val preprocessingRules = Seq(
+          planSubqueriesRule)
+        // Run pre-processing rules.
+        val newPlan = AdaptiveSparkPlanExec.applyPhysicalRules(plan, preprocessingRules)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala]: can we do it at https://github.com/apache/spark/pull/25471/files#diff-6954dd8020a9ca298f1fb9602c0e831cR125 ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala (hunk 2 above)
Comment (maryannxue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: We have split the pre-planning rules into two groups now: `preprocessingRules` and `queryStagePreparationRules`, the former is necessary before the creation of `AdaptiveSparkPlanExec` because otherwise the initial physical plan will have logical subquery nodes in it. The latter, though, can be deferred to after `AdaptiveSparkPlanExec` creation.

**Comment by maryannxue** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+package org.apache.spark.sql.execution.adaptive
+
+import scala.collection.concurrent.TrieMap
+
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{BaseSubqueryExec, ExecSubqueryExpression, ReusedSubqueryExec, SparkPlan}
+import org.apache.spark.sql.internal.SQLConf
+
+case class ReuseAdaptiveSubquery(
+  conf: SQLConf,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala]: nit: 4 space indentation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala]: Can we do it when assigning the default value of `currentPhysicalPlan`? e.g.
```
@volatile private var currentPhysicalPlan = applyPhysicalRules(initialPlan, queryStagePreparationRules)
```

**Comment by maryannxue** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by dilipbiswal** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!
